StringCalculatorKata/ILoopAction.cs

```
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace StringCalculatorKata
+{
+    public interface ILoopAction
+    {
+        public void DoAction(ref int currentElement);
+    }
+
+    public class SummationLoopAction : ILoopAction
+    {
+        public int Summation { get; set; } = 0;
+
+        public void DoAction(ref int currentElement)
+        {
+            Summation += currentElement;
+        }
+    }
+
+    public class CheckForNegativesLoopAction : ILoopAction
+    {
+        private readonly StringCalculator _stringCalculator;
+        private readonly List<int> _negativeNumbers;
+        private const string ExceptionMessageIntro = "negatives are not allowed";
+
+        public CheckForNegativesLoopAction(StringCalculator stringCalculator)
+        {
+            _stringCalculator = stringCalculator;
+            _negativeNumbers = new List<int>();
+            stringCalculator.DoneAddLoop +=
+                () =>
+                {
+                    if (_negativeNumbers.Count > 0)
+                        throw new Exception($"{ExceptionMessageIntro} : {string.Join(", ", _negativeNumbers)}");
```

you can throw InvalidOperationException as well with this message.